src/dmd/dmodule.d

```diff
@@ -662,7 +662,27 @@ extern (C++) final class Module : Package
             return true; // already read
 
         //printf("Module::read('%s') file '%s'\n", toChars(), srcfile.toChars());
-        if (auto result = global.fileManager.lookup(srcfile))
+
+        /* Preprocess the file if it's a C file
+         */
+        FileName filename = srcfile;
+        bool ifile = false;             // did we generate a .i file
+        scope (exit)
+        {
+            if (ifile)
+                File.remove(filename.toChars());        // remove generated file
+        }
+        version (linux)
+        {
+            if (FileName.equalsExt(srcfile.toString(), c_ext) &&
+                FileName.exists(srcfile.toString()))
+            {
+                filename = global.preprocess(srcfile);  // run C preprocessor
```

Comment:
Check for `!is null`.

Reply:
It returns a struct, not a pointer.

Reply:
I presume he means `if (global.preprocess !is null) filename = global.preprocess(srcfile);`?

Reply:
> I prefer to ask @ibuclaw to explain, as I tire of guessing :-)

Yes, the function pointer might be null.

Reply:
> I expect a "no-op preprocess" function would be provided that simply returned its argument. It's more attractive than adding the extra test.

If you want that, `global.preprocess` should have a "contract" check for non-null function pointers. I agree with @ibuclaw anyway, because that way, there is a way to disable preprocess for those files.

Reply:
Don't need a contract check for null pointers, as the hardware does that for you.

Reply:
> it's counter-intuitive

It's a different style, called branchless programming. The nice thing about it is it removes clutter from the code.

> and a function call has higher latency for that purpose, instead of a simple test check.

I'd agree if it was an inner loop, but it's once per file. Not significant.

Reply:
Initially this would be unimplemented in gdc and ldc though, resulting in a segfault. If this field was default initialized with a noop lambda then that'd be different.